includes/WcfceMailProcessor.inc

```diff
@@ -416,6 +416,84 @@ class WcfceMailProcessor {
    *   An exception generated by civicrm_api3().
    */
   private function getContactIdByEmail($email) {
+    $res = $this->getContactIdByEmailFromWebform($email);
+
+    if (!$res) {
+      $res = $this->getContactIdByEmailFromDb($email);
+    }
+
+    return $res;
+  }
+
+  /**
+   * Set contact id by the value of some email webform field.
+   *
+   * If you have multiple contacts with the same email, there could be a
+   * confusion: as contacts are retrieved by email when activity is created,
+   * some other contact may be accidentally used.
+   * However, as yo may know it is possible to configure webform notification to
+   * get From/To email address from some form field. If this field comes from
+   * civicrm contact that you've configured under the Civicrm tab of the form
+   * then it is possible to link email address to a contact, we just need to
+   * find that field and parse it's name (form key). So that's what this method
+   * actually does.
+   * Keep in mind: in order to make it work for hidden fields you need to make
+   * sure it's properly configured: go to field settings and check if
+   * 'Hidden type' is set to 'Hidden element' (not 'Secure value').
+   *
+   * @param string $email
+   *   Email address to search by.
+   *
+   * @return int
+   *   Civicrm contact id, or 0 if not found.
+   */
+  private function getContactIdByEmailFromWebform($email) {
+    $contactId = 0;
+    $fields = $this->message['params']['node']->webform['components'];
+    $values = $this->message['params']['submission']->data;
+    $civicrm_data = $this->message['params']['submission']->civicrm;
+
+    // Go through all submitted data.
+    foreach ($values as $num => $value) {
+      $fieldKey = '';
+      $contactNum = '';
+
+      // If a field contains email address that we're looking for - get field
+      // key (name).
+      if (trim($value[0]) === $email) {
+        $fieldKey = $fields[$num]['form_key'] ?? '';
+      }
+
+      // If field key looks like civicrm email field - extract contact number
+      // from it (key would look like 'civicrm_3_contact_2_email_email').
+      if ($fieldKey) {
+        preg_match('!civicrm_(\d+)_contact_\d+_email!', $fieldKey, $match);
+        $contactNum = $match[1] ?? '';
+      }
+
+      // Webform stores contacts used in a form in array, so we can get
+      // contact id from here.
+      if ($contactNum) {
+        $contactId = $civicrm_data['contact'][$contactNum]['id'] ?? 0;
```

Comment:
We should do a quick return once the contact Id is found instead of looping through all data in `$values`. Why don't we do a break here and exit the loop.